File: common/src/main/java/org/tdf/common/store/StoreWrapper.java
package org.tdf.common.store;

import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NonNull;
import org.tdf.common.serialize.Codec;

import java.util.AbstractMap;
import java.util.Collection;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * delegate {@code Store<U, R>} as {@code Store<K, V>}
 *
 * @param <K> type of key
 * @param <V> type of value
 */
@AllArgsConstructor
public class StoreWrapper<K, V, U, R>
        implements BatchStore<K, V> {

    @Getter
    private Store<byte[], byte[]> store;

    private Codec<K> keyCodec;
    private Codec<V> valueCodec;


    @Override
    public V get(@NonNull K k) {
        byte[] v = store.get(keyCodec.getEncoder().apply(k));
        if (v == null || v.length == 0)
            return null;
        return valueCodec.getDecoder().apply(v);
    }

    @Override
    public void put(@NonNull K k, @NonNull V v) {
        store.put(keyCodec.getEncoder().apply(k), valueCodec.getEncoder().apply(v));
    }


    @Override
    public void remove(@NonNull K k) {
        store.remove(keyCodec.getEncoder().apply(k));
    }

    @Override
    public void flush() {

    }


    @Override
    public void putAll(Collection<? extends Map.Entry<? extends K, ? extends V>> rows) {
        ((BatchStore<byte[], byte[]>) store).putAll(
                rows.stream().map(e -> new AbstractMap.SimpleEntry<>(
                        keyCodec.getEncoder().apply(e.getKey()),
                        valueCodec.getEncoder().apply(e.getValue())
                ))
                        .collect(Collectors.toList())
        );
    }
}

File: common/src/main/java/org/tdf/common/store/StoreWrapper.kt
package org.tdf.common.store

import org.tdf.common.serialize.Codec

/**
 * delegate `Store<ByteArray, ByteArray>` as `Store<K, V>`
 *
 * @param <K> type of key
 * @param <V> type of value
</V></K> */
class StoreWrapper<K, V>(
    val store: Store<ByteArray, ByteArray>,
    private val keyCodec: Codec<K>,
    private val valueCodec: Codec<V>
) : Store<K, V> {
    override fun get(k: K): V? {
        val v = store[keyCodec.encoder.apply(k)]
        return if (v == null || v.isEmpty()) null else valueCodec.decoder.apply(v)
    }

    override fun set(k: K, v: V) {
        store[keyCodec.encoder.apply(k)] = valueCodec.encoder.apply(v)
    }

    override fun remove(k: K) {
        store.remove(keyCodec.encoder.apply(k))
    }

    override fun flush() {}
}